tla/etcdraft.tla

```diff
@@ -0,0 +1,913 @@
+       \/ /\ \lnot m.msuccess \* not successful
+          /\ UNCHANGED <<leaderVars>>
+    /\ Discard(m)
+    /\ UNCHANGED <<pendingMessages, serverVars, candidateVars, logVars, configVars, durableState>>
+
+\* Any RPC with a newer term causes the recipient to advance its term first.
+\* @type: (Int, Int, MSG) => Bool;
+UpdateTerm(i, j, m) ==
+    /\ m.mterm > currentTerm[i]
+    /\ BecomeFollowerOfTerm(i, m.mterm)
+       \* messages is unchanged so m can be processed further.
+    /\ UNCHANGED <<messageVars, candidateVars, leaderVars, logVars, configVars, durableState>>
+
+\* Responses with stale terms are ignored.
+\* @type: (Int, Int, MSG) => Bool;
+DropStaleResponse(i, j, m) ==
+    /\ m.mterm < currentTerm[i]
+    /\ Discard(m)
+    /\ UNCHANGED <<pendingMessages, serverVars, candidateVars, leaderVars, logVars, configVars, durableState>>
+
+\* Receive a message.
+ReceiveDirect(m) ==
+    LET i == m.mdest
+        j == m.msource
+    IN \* Any RPC with a newer term causes the recipient to advance
+       \* its term first. Responses with stale terms are ignored.
+    \/ UpdateTerm(i, j, m)
+    \/  /\ m.mtype = RequestVoteRequest
+        /\ HandleRequestVoteRequest(i, j, m)
+    \/  /\ m.mtype = RequestVoteResponse
+        /\  \/ DropStaleResponse(i, j, m)
+            \/ HandleRequestVoteResponse(i, j, m)
+    \/  /\ m.mtype = AppendEntriesRequest
+        /\ HandleAppendEntriesRequest(i, j, m)
+    \/  /\ m.mtype = AppendEntriesResponse
+        /\ \/ DropStaleResponse(i, j, m)
+           \/ HandleAppendEntriesResponse(i, j, m)
+
+Receive(m) == ReceiveDirect(m)
+
+NextRequestVoteRequest == \E m \in DOMAIN messages : m.mtype = RequestVoteRequest /\ Receive(m)
+NextRequestVoteResponse == \E m \in DOMAIN messages : m.mtype = RequestVoteResponse /\ Receive(m)
+NextAppendEntriesRequest == \E m \in DOMAIN messages : m.mtype = AppendEntriesRequest /\ Receive(m)
+NextAppendEntriesResponse == \E m \in DOMAIN messages : m.mtype = AppendEntriesResponse /\ Receive(m)
+
+\* End of message handlers.
+----
+\* Network state transitions
+
+\* The network duplicates a message
+\* @type: MSG => Bool;
+DuplicateMessage(m) ==
+    /\ m \in DOMAIN messages
+    /\ messages' = WithMessage(m, messages)
+    /\ UNCHANGED <<pendingMessages, serverVars, candidateVars, leaderVars, logVars, configVars, durableState>>
+
+\* The network drops a message
+\* @type: MSG => Bool;
+DropMessage(m) ==
+    \* Do not drop loopback messages
+    \* /\ m.msource /= m.mdest
+    /\ Discard(m)
+    /\ UNCHANGED <<pendingMessages, serverVars, candidateVars, leaderVars, logVars, configVars, durableState>>
+
+----
+
+\* Defines how the variables may transition.
+NextAsync == 
+    \/ \E i,j \in Server : RequestVote(i, j)
+    \/ \E i \in Server : BecomeLeader(i)
+    \/ \E i \in Server: ClientRequest(i, 0)
+    \/ \E i \in Server : AdvanceCommitIndex(i)
+    \/ \E i,j \in Server : \E b,e \in matchIndex[i][j]+1..Len(log[i])+1 : AppendEntries(i, j, <<b,e>>)
```

Comment:
Change possible entry ranges to be replicated to respect the fact that next index is always larger than match index.